connector/signaltometricsconnector/internal/aggregator/aggregator.go

```diff
@@ -0,0 +1,336 @@
+// Licensed to Elasticsearch B.V. under one or more contributor
+// license agreements. See the NOTICE file distributed with
+// this work for additional information regarding copyright
+// ownership. Elasticsearch B.V. licenses this file to you under
+// the Apache License, Version 2.0 (the "License"); you may
+// not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+package aggregator // import "github.com/elastic/opentelemetry-collector-components/connector/signaltometricsconnector/internal/aggregator"
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"time"
+
+	"github.com/elastic/opentelemetry-collector-components/connector/signaltometricsconnector/internal/metadata"
+	"github.com/elastic/opentelemetry-collector-components/connector/signaltometricsconnector/internal/model"
+	"github.com/open-telemetry/opentelemetry-collector-contrib/pkg/ottl"
+	"github.com/open-telemetry/opentelemetry-collector-contrib/pkg/pdatautil"
+	"go.opentelemetry.io/collector/pdata/pcommon"
+	"go.opentelemetry.io/collector/pdata/pmetric"
+)
+
+// Aggregator provides a single interface to update all metrics
+// datastructures. The required datastructure is selected using
+// the metric definition.
+type Aggregator[K any] struct {
+	result pmetric.Metrics
+	// smLookup maps resourceID against scope metrics since the aggregator
+	// always produces a single scope.
+	smLookup    map[[16]byte]pmetric.ScopeMetrics
+	valueCounts map[model.MetricKey]map[[16]byte]map[[16]byte]*valueCountDP
+	sums        map[model.MetricKey]map[[16]byte]map[[16]byte]*sumDP
+	timestamp   time.Time
+}
+
+// NewAggregator creates a new instance of aggregator.
+func NewAggregator[K any](metrics pmetric.Metrics) *Aggregator[K] {
```

Comment:
[**For reviewers**] This is currently missing unit-tests and is only tested via the connector tests, I will add these as a follow-up.